### PR TITLE
Add protocol to do serialisation/deserialisation

### PR DESCRIPTION
**Pull request by foolswood**

### CODE REVIEW

File: clapi-hs.cabal [hunk 1]
```diff
@@ -18,6 +18,7 @@ library
   exposed-modules:     Data.Map.Mos
                      , Data.Map.Mol
                      , Data.Map.Clapi
+                     , Clapi.EncodingProtocol
```
Comment (ch3pjw) [inline clapi-hs.cabal]: `CodecProtocol`?

File: test/TestEncodingProtocol.hs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import Clapi.EncodingProtocol (serialiser)
+
+tests = [
+    testCase "packetised round trip" testPacketisedRoundTrip
+    ]
+
+testPacketisedRoundTrip = do
+    (bi, bo) <- U.newChan
+    (mi, mo) <- U.newChan
+    mv <- newEmptyMVar
```
Comment (ch3pjw) [inline test/TestEncodingProtocol.hs]: I'm not sure I follow what this `MVar` is for. Could you do any indefinitely blocking operation where it's used on the next line?

Comment (foolswood) [reply test/TestEncodingProtocol.hs]: I think so yes, just didn't know how to do it any other way.

File: clapi-hs.cabal (hunk 1 above)
Comment (foolswood) [reply clapi-hs.cabal]: To me that would imply a stream, but maybe this is SerialisationProtocol or something as it is the join of those 2 things?

File: src/Clapi/EncodingProtocol.hs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+serialiser :: Monad m => Protocol ByteString [Message] ByteString [Message] m ()
+serialiser = serialiser' $ parse parser
+  where
+    serialiser' p = waitThen (fwd p) (rev p)
+    fwd parseNext bs = case parseNext bs of
+        Fail _ ctxs err -> sendRev $ fromString err
+        Partial cont -> serialiser' cont
+        Done unconsumed msgs -> sendFwd msgs >> fwd (parse parser) unconsumed
+    rev p msgs = either (error "encode failed") (\bs -> sendRev bs >> serialiser' p) (encode msgs)
```
[the remark this answers is not in the document]
Comment (foolswood) [reply src/Clapi/EncodingProtocol.hs]: Sure, why?

**Pull request merged**